### DISCUSSION TITLE
Generating an image from the map canvas results in blank images.

### DISCUSSION THREAD

**Discussion by daamsie** [maplibre/maplibre-gl-js#3900] [Q&A]
**maplibre-gl-js version**: latest

**browser**: Chrome

### Steps to Trigger Behavior

 1. set the pixelRatio of the map using `map.setPixelRatio()` to a higher resolution.
 2. set up a resize handler that will generate an image of the map using `map.getCanvas().toDataURL()` or `map.getCanvas().toBlob()` 
 3. trigger `map.resize()`

### Link to Demonstration
https://output.jsbin.com/tikuhuk/5

### Expected Behavior
The image saves at the higher resolution.

### Actual Behavior
It _almost always_ saves a blank image in Chrome. 

Opening the dev inspector while saving the image does result in the save working, which makes me think there is some kind of timing issue going on that is affected by the inspector being open

My research into this issue led me to the preserveDrawingBuffer setting which seemed like it must be somehow related. However, setting this to true in the maplibre options does not resolve the issue. 

Relevant code below because jsbin is not working properly

```
<script>
  var map = new maplibregl.Map({
    container: 'map',
    style: 'https://demotiles.maplibre.org/style.json',
    center: [0, 0],
    zoom: 1,
    preserveDrawingBuffer: true
  });
  
  function saveImage() {
    // Set the DPI to higher res
    map.setPixelRatio(300 / 48);

    // on resize save the image
    map.once('resize', function () {
      var link = document.createElement('a');
      link.href = map.getCanvas().toDataURL("image/png")
      link.download = 'map.png'
      document.body.appendChild(link);
      link.click();
      document.body.removeChild(link);
    });

    map.resize();
  }
</script>
```

**Comment by sbachinin** [comment]
I think using resize event isn't a good idea.
Try "idle" instead. For me this works:
```
function saveImage() {
    map.setPixelRatio(300 / 48);

    map.once('idle', () => {
        const link = document.createElement('a');
        link.href = map.getCanvas().toDataURL('image/png');
        link.download = 'map.png';
        document.body.appendChild(link);
        link.click();
    });
}
```

**Comment by daamsie** [comment]
Thanks for the response!

I have tried any number of combos with this and had the same problem with all of them. 

But you're quite right - in the jsbin the "idle" event is indeed working ok.

I have tested the 'idle' listener in my own project again and it's still producing blank images. My theory here is similar to the difference between having the developer tools open vs not. The timing is just slightly different for the idle handler in the pastebin vs in my project which of course is drawing a lot more on the map than just this simple example.

**Comment by HarelM** [comment]
You need to use preserveDrawingBuffer.
The link to jsbin isn't working.

**Reply by daamsie** [reply]
I am using preserveDrawingBuffer. 

I have added the jsbin code in the question because jsbin is not working.

**Comment by sbachinin** [comment]
As far as I know, preserveDrawingBuffer was already used in this example.
@daamsie, will this work in your project?
```
function saveImage() {
    map.setPixelRatio(300 / 48);
    setTimeout(() => {
        const link = document.createElement('a');
        link.href = map.getCanvas().toDataURL('image/png');
        link.download = 'map.png';
        document.body.appendChild(link);
        link.click();
        document.body.removeChild(link);
    });
}
```

**Reply by daamsie** [reply]
I've given this a try too with no success. I should say - this was all working fairly reliably until a few months ago. 

Something updated in Chrome and has caused it to start breaking. I'm having a real hard time working out where the issue is.

**Reply by HarelM** [reply]
You can look at the ngx-maplibre-gl cypress tests to see how we compare images or the generate image script in this repo. All of them areable to extract an image from the canvas.

**Reply by daamsie** [reply]
Thanks for that. It looks like the same basic principle, but I couldn't find any examples where the pixel ratio is being changed prior to saving the screenshot. 

If I just capture a screenshot without changing the pixel ratio first there are no issues. I will try and get another pastebin up with some code that demonstrates the problem.

**Comment by daamsie** [comment]
I seem to have a fix. In my stupidity I hadn't updated the maplibre library as my first step. I assumed that the problem existed in the current version because my jsbin test also had the issue. That however was more down to using the 'resize' event handler rather than the 'idle' handler. 

After updating to the latest version I am able to quite reliably export images at a high res now.

I even added in an extra step to detect the user's max webgl texture size to push the size past the 4096 pixels. It's generating some very nice high resolution maps.

Thanks to @HarelM and @sbachinin for taking the time to respond and provide assistance.
